### PR TITLE
feat(inputs.cisco_telemetry_mdt): Include delete field

### LINKED ISSUE #12286: Cisco Model-Driven Telemetry (MDT) Input Plugin: Doesn't support/recognize Event-Driven Telemetry (EDT) Delete Messages
### Relevant telegraf.conf

```toml
[agent]
    interval = "10s"
    round_interval = true
    metric_buffer_limit = 10000
    flush_buffer_when_full = true
    collection_jitter = "0s"
    flush_interval = "10s"
    flush_jitter = "0s"
    debug = false
    quiet = false
    hostname = "telegraf"
[[inputs.cisco_telemetry_mdt]]
    transport = "grpc"
    service_address = ":32400"
[[outputs.kafka]]
    brokers = ["kafka:9092"]
    topic = "test.telemetry"
    data_format = "json"
[[outputs.file]]
    files = ["stdout"]
    data_format = "json"
```

### Logs from Telegraf

```text
# add ipv6 address on lo10
{"fields":{"state/ip":"10::10","state/origin":"STATIC","state/prefix_length":128,"state/status":"PREFERRED"},"name":"openconfig-interfaces:interfaces/interface/subinterfaces/subinterface/openconfig-if-ip:ipv6/addresses/address","tags":{"host":"telegraf","index":"0","ip":"10::10","name":"Loopback10","path":"openconfig-interfaces:interfaces/interface/subinterfaces/subinterface/openconfig-if-ip:ipv6/addresses/address","source":"sfo","subscription":"base_metrics"},"timestamp":1669633122}

# remove ipv6 on lo10 (this message is not shown on newer telemetry versions - it was removed in https://github.com/influxdata/telegraf/commit/dbd1fe734ba9222999cac2ad3cab5baf4023e6f2)
2022-11-28T10:58:27Z I! [inputs.cisco_telemetry_mdt] Message from sfo missing keys or content
```

### System info

Telegraf 1.15-alpine (Docker), Ubuntu 22.04.1 LTS

### Docker

The problem is not related to docker - anyway here is the `docker-compose.yml`: 
```
version: "3"
services:
  zookeeper:
    image: 'bitnami/zookeeper:latest'
    ports:
      - '2181:2181'
    environment:
      - ALLOW_ANONYMOUS_LOGIN=yes
  kafka:
    image: 'bitnami/kafka:latest'
    ports:
      - '9092:9092'
      - '9093:9093'
    environment:
      - KAFKA_BROKER_ID=1
      - KAFKA_CFG_ZOOKEEPER_CONNECT=zookeeper:2181
      - KAFKA_CFG_LISTENER_SECURITY_PROTOCOL_MAP=CLIENT:PLAINTEXT,EXTERNAL:PLAINTEXT
      - KAFKA_CFG_LISTENERS=CLIENT://:9092,EXTERNAL://:9093
      - KAFKA_CFG_ADVERTISED_LISTENERS=CLIENT://kafka:9092,EXTERNAL://172.16.16.24:9093
      - KAFKA_CFG_INTER_BROKER_LISTENER_NAME=CLIENT
      - ALLOW_PLAINTEXT_LISTENER=yes
    depends_on:
      - zookeeper
  telegraf:
    image: 'telegraf:1.15-alpine'
    ports:
      - '32400:32400'
    depends_on:
      - zookeeper
      - kafka

    volumes:
      - "./telegraf.conf:/etc/telegraf/telegraf.conf:ro"
```

Just to notice: the bug is also present in more recent telegraf version. But there the log message was removed, so we had to use the older version `telegraf:1.15-alpine`. For more details look at: https://github.com/influxdata/telegraf/commit/dbd1fe734ba9222999cac2ad3cab5baf4023e6f2.

### Steps to reproduce

1. Configure the cisco device with event-driven telemetry (on IOS-XR sample-interval 0):
Here is an example with IOS-XR:
```
telemetry model-driven
 destination-group test
  address-family ipv4 xxx.xxx.xxx.xxx port 32400
   encoding self-describing-gpb
   protocol grpc no-tls
  !
 !
 sensor-group openconfig
  sensor-path openconfig-interfaces:interfaces/interface/subinterfaces/subinterface/openconfig-if-ip:ipv6/addresses/address
 !
 subscription base_metrics
 ! here the event-driven configuration (sample-interval 0)
 ! more info: https://www.cisco.com/c/en/us/td/docs/iosxr/ncs5500/telemetry/63x/b-telemetry-cg-ncs5500-63x/b-telemetry-cg-ncs5500-63x_chapter_010.html#id_36274
  sensor-group-id openconfig sample-interval 0  
  destination-id test
 !
!

```
2. On the router configure an example interface with ipv6:
```
RP/0/RP0/CPU0:sfo(config)#interface lo10 ipv6 address 10::10/128
RP/0/RP0/CPU0:sfo(config)#commit

``` 
3. Remove the ipv6 address of the interface with 
```
RP/0/RP0/CPU0:sfo(config)#no interface lo10 ipv6 address 10::10/128
RP/0/RP0/CPU0:sfo(config)#commit

```
5.  Observe the telemetry log message (only on older versions):
```
2022-11-28T11:10:52Z I! [inputs.cisco_telemetry_mdt] Message from sfo missing keys or content
```
...

### Expected behavior

Telegraf should process the deleted message and create a metric with a name, tags and fields (possibly empty objects) and forward it to the configured output.

### Actual behavior

The delete message from event-driven telemetry (EDT) is not processed at all. The logic is not implemented: https://github.com/influxdata/telegraf/blob/c3562ae8b27cb9b382ec86d41e183156f3f18f3d/plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go#L391

### Additional info

_No response_

### PR DESCRIPTION
**Pull request by severindellsperger**
# Required for all PRs

- [x] Updated associated README.md.
- [x] Wrote appropriate unit tests.
- [x] Pull request title or commits are in [conventional commit format](https://www.conventionalcommits.org/en/v1.0.0/#summary)

resolves #12286

Implement the logic to include the `delete` field as discussed in #12286 to support Event-Driven Telemetry (EDT).

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -19,6 +19,7 @@ require (
 	github.com/BurntSushi/toml v1.2.1
 	github.com/ClickHouse/clickhouse-go v1.5.4
 	github.com/DATA-DOG/go-sqlmock v1.5.0
+	github.com/INSRapperswil/cisco-xr-telemetry-proto v1.0.1
```
Comment (srebhan) [inline go.mod:22]: Can you please try to file a PR against https://github.com/cisco-ie/nx-telemetry-proto to add your generated code!? It would be nice to not have to randomly switch repositories for this rather critical part...

File: plugins/inputs/cisco_telemetry_mdt/sample.conf [hunk 2]
```diff
@@ -50,3 +53,5 @@
   ## GRPC minimum timeout between successive pings, decreasing this value may 
   ## help if this plugin is closing connections with ENHANCE_YOUR_CALM (too_many_pings).
   # keepalive_minimum_time = "5m"
+
+
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/sample.conf:57]: Please only add one newline if at all.

File: plugins/inputs/cisco_telemetry_mdt/sample.conf [hunk 3]
```diff
@@ -21,6 +21,9 @@
  ## Define (for certain nested telemetry measurements with embedded tags) which fields are tags
  # embedded_tags = ["Cisco-IOS-XR-qos-ma-oper:qos/interface-table/interface/input/service-policy-names/service-policy-instance/statistics/class-stats/class-name"]
 
+## Include the delete field in every telemetry message.
+  include_delete_field = true
+
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/sample.conf:26]: This is not correct. The default should keep backward compatibility. Furthermore, this is not an option the user _has to_ set so I'd rather like to see this as a (commented) optional param with the default:
```suggestion
  ## Include the delete field in every telemetry message.
  # include_delete_field = false
```

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 4]
```diff
@@ -51,13 +51,14 @@ type GRPCEnforcementPolicy struct {
...
 type CiscoTelemetryMDT struct {
 	// Common configuration
-	Transport         string
-	ServiceAddress    string                `toml:"service_address"`
-	MaxMsgSize        int                   `toml:"max_msg_size"`
-	Aliases           map[string]string     `toml:"aliases"`
-	Dmes              map[string]string     `toml:"dmes"`
-	EmbeddedTags      []string              `toml:"embedded_tags"`
-	EnforcementPolicy GRPCEnforcementPolicy `toml:"grpc_enforcement_policy"`
+	Transport          string
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:54]: When we are at it, can you please also add the (unrelated) missing TOML annotation here?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 5]
```diff
@@ -398,16 +398,20 @@ func (c *CiscoTelemetryMDT) handleTelemetry(data []byte) {
 		if msgID := msg.GetSubscriptionIdStr(); msgID != "" {
 			tags["subscription"] = msgID
 		}
-		tags["path"] = msg.GetEncodingPath()
+		encodingPath := msg.GetEncodingPath()
+		tags["path"] = encodingPath
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:402]: Why did you change this?

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 6]
```diff
@@ -398,16 +398,20 @@ func (c *CiscoTelemetryMDT) handleTelemetry(data []byte) {
...
 
-		// Parse values
-		for _, subfield := range content.Fields {
-			c.parseContentField(grouper, subfield, "", msg.EncodingPath, tags, timestamp)
+		if content != nil {
+			// Parse values
+			for _, subfield := range content.Fields {
+				c.parseContentField(grouper, subfield, "", encodingPath, tags, timestamp)
+			}
 		}
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:413]: How about
```suggestion
		if content == nil {
			continue
		}

		// Parse values
		for _, subfield := range content.Fields {
			c.parseContentField(grouper, subfield, "", encodingPath, tags, timestamp)
		}
```
as this will save an indentation level and reduces the patch size.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go [hunk 7]
```diff
@@ -96,6 +96,166 @@ func TestHandleTelemetryTwoSimple(t *testing.T) {
...
+	for _, test := range testCases {
+		c := &CiscoTelemetryMDT{
+			Log:                testutil.Logger{},
+			Transport:          "dummy",
+			Aliases:            map[string]string{"deleted": encodingPath.stringValue},
+			IncludeDeleteField: true}
+		acc := &testutil.Accumulator{}
+		err := c.Start(acc)
+		// error is expected since we are passing in dummy transport
+		require.Error(t, err)
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go:249]: Can we please reflect this in the `require` statement
```suggestion
		// error is expected since we are passing in dummy transport
		require.ErrorContains(t, c.Start(acc), "a message part you expect")
```

File: go.mod (hunk 1 above)
Comment (severindellsperger) [reply go.mod:22]: The mentioned repo is for Cisco NXOS and not IOS-XR, so I think a PR would be declined anyway.
Unfortunately, the only +/- current proto file including the delete field is for [IOS-XR](https://github.com/ios-xr/model-driven-telemetry) (working also for NXOS). Cisco doesn't provide the Go implementation for this proto file -> this is why we have created this repo. An advantage is that we can at least react to updated proto files fast.

Another option would be to save the proto files directly in the telegraf repo, but I think this was not intended to do.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 8]
```diff
@@ -386,9 +387,8 @@ func (c *CiscoTelemetryMDT) handleTelemetry(data []byte) {
 			}
 		}
 
-		// if the keys and content fields are missing, skip the message as it
-		// does not have parsable data used by Telegraf
-		if keys == nil || content == nil {
+		if content == nil && !c.IncludeDeleteField {
+			c.Log.Debugf("Message skipped because no content found and include of delete field not enabled")
```
[the remark this answers is not in the document]
Comment (severindellsperger) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:391]: There are models, where the `keys` can be `nil` and the `content` not. An example is `Cisco-IOS-XR-fib-common-oper:oc-aft-l3/protocol/ipv4/oor/state` (see also [here](https://www.cisco.com/c/en/us/td/docs/iosxr/ncs5500/telemetry/75x/b-telemetry-cg-ncs5500-75x/enhancememts-to-telemetry.html#Cisco_Concept.dita_7912af05-c8d8-40af-98de-c87be377e754)). 
The code panicked when only the `keys` were nil and the `content` wasn't (because you checked only if both (`keys` and `content` is `nil`)
I fixed this behavior in 9aabe02
keys_nil

PS: The metric is still valid because there are other tags added based on XPath, etc.

File: plugins/inputs/cisco_telemetry_mdt/sample.conf (hunk 2 above)
Comment (severindellsperger) [reply plugins/inputs/cisco_telemetry_mdt/sample.conf]: Solved in [8123f7e](https://github.com/influxdata/telegraf/pull/12345/commits/8123f7e46f4f13847d1020c12fcb398ea1f13312) respectively with update of the docs in [36355d8](https://github.com/influxdata/telegraf/pull/12345/commits/36355d8c9cb9442ffbf4aeb21efde42d82578a64)

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 9]
```diff
@@ -15,8 +15,8 @@ import (
 	"sync"
 	"time"
 
+	telemetry "github.com/INSRapperswil/cisco-xr-telemetry-proto"
 	dialout "github.com/cisco-ie/nx-telemetry-proto/mdt_dialout"
-	telemetry "github.com/cisco-ie/nx-telemetry-proto/telemetry_bis"
```
[the remark this answers is not in the document]
Comment (severindellsperger) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go]: Done: faac53a

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go (hunk 4 above)
Comment (severindellsperger) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go]: Done: [8fe6d33](https://github.com/influxdata/telegraf/pull/12345/commits/8fe6d33ef5ad0fe4f4f5911b4cdbf447bd80cc46)

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go (hunk 6 above)
Comment (severindellsperger) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go]: Your proposal would lead that the delete field would not be included. I solved it by adding the delete field (if enabled) before parsing the values. See: [faac53a](https://github.com/influxdata/telegraf/pull/12345/commits/faac53a914700b4bfe816acc4babe21a2055a7b4)

File: go.mod (hunk 1 above)
Comment (severindellsperger) [reply go.mod:22]: Update: requested cisco to create its own repo - waiting for an answer.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (severindellsperger) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go]: The pipeline does not work because `gofmt` does not accept the order as proposed...

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go [hunk 10]
```diff
@@ -96,6 +96,166 @@ func TestHandleTelemetryTwoSimple(t *testing.T) {
...
+	name := TelemetryEntry{name: "name", stringValue: "Loopback10"}
+	index := TelemetryEntry{name: "index", stringValue: "0"}
+	ip := TelemetryEntry{name: "ip", fieldName: "state/ip", stringValue: "10::10"}
+	prefixLength := TelemetryEntry{name: "prefix-length", fieldName: "state/prefix_length", uint32Value: uint32(128), uint64Value: 128}
+	origin := TelemetryEntry{name: "origin", fieldName: "state/origin", stringValue: "STATIC"}
+	status := TelemetryEntry{name: "status", fieldName: "state/status", stringValue: "PREFERRED"}
+	source := TelemetryEntry{name: "source", stringValue: "hostname"}
+	subscription := TelemetryEntry{name: "subscription", stringValue: "subscription"}
+	deleteKey := "delete"
+	stateKey := "state"
```
[the remark this answers is not in the document]
Comment (severindellsperger) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go:120]: I don't understand: they are in the test case. However, these elements are used for all subtests. This is why I defined these here.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go [hunk 11]
```diff
@@ -96,6 +96,166 @@ func TestHandleTelemetryTwoSimple(t *testing.T) {
...
+			source.name:       source.stringValue,
+			subscription.name: subscription.stringValue,
+		},
+		fields: map[string]interface{}{
+			deleteKey:              false,
+			ip.fieldName:           ip.stringValue,
+			prefixLength.fieldName: prefixLength.uint64Value,
+			origin.fieldName:       origin.stringValue,
+			status.fieldName:       status.stringValue,
+		},
```
[the remark this answers is not in the document]
Comment (severindellsperger) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go:196]: Indeed, I could introduce an `expected`  and test it with `testutil.RequireMetricsEqual`. 
I just don't see how I can the `actual telegraf.Metric`, to compare the actual with expected Metric(s).
Can you explain?

File: plugins/inputs/cisco_telemetry_mdt/sample.conf [hunk 12]
```diff
@@ -49,4 +51,4 @@
 
   ## GRPC minimum timeout between successive pings, decreasing this value may 
   ## help if this plugin is closing connections with ENHANCE_YOUR_CALM (too_many_pings).
-  # keepalive_minimum_time = "5m"
+  # keepalive_minimum_time = "5m"
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/sample.conf:54]: Can you please keep the trailing newline?

File: plugins/inputs/cisco_telemetry_mdt/sample.conf [hunk 13]
```diff
@@ -21,6 +21,8 @@
  ## Define (for certain nested telemetry measurements with embedded tags) which fields are tags
  # embedded_tags = ["Cisco-IOS-XR-qos-ma-oper:qos/interface-table/interface/input/service-policy-names/service-policy-instance/statistics/class-stats/class-name"]
 
+  ## Include the delete field in every telemetry message.
+  # include_delete_field = false
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/sample.conf:25]: Maybe add a newline to separate this option from the rest of the config
```suggestion
  ## Include the delete field in every telemetry message.
  # include_delete_field = false
  
```

**Review by srebhan — APPROVED** [review]
Thanks for your explanations @severindellsperger! I have two cosmetic suggestions for the `sample.conf` but nothing to hold up this PR.

@powersj what do you think about the dependency? Should we merge as is and adapt later once there is an "official Cisco release" or should we go with the other repo?

**Comment by severindellsperger** [comment]
> @severindellsperger, did you ask Cisco to host the repo publicly? Have they acknowledged the request?
I have no answer yet. I have a meeting next week and will discuss it then.

File: go.mod (hunk 1 above)
Comment (severindellsperger) [reply go.mod:22]: @srebhan @powersj FYI: Finally, I got the clearance to do a PR on the nx-telemetry-proto repo with the latest files. I will do it tomorrow and change the import path in this PR.

**Comment by severindellsperger** [comment]
The PR is out: https://github.com/cisco-ie/nx-telemetry-proto/pull/2
Let's hope it will be merged soon so that we can adjust.

**Comment by powersj** [comment]
> The PR is out: [cisco-ie/nx-telemetry-proto#2](https://github.com/cisco-ie/nx-telemetry-proto/pull/2) Let's hope it will be merged soon so that we can adjust.

Looks like that got landed. Thank you for doing that work!

**Comment by severindellsperger** [comment]
@powersj  @srebhan I made the necessary changes. I think the PR is now ready for the final review.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go [hunk 14]
```diff
@@ -386,27 +387,38 @@ func (c *CiscoTelemetryMDT) handleTelemetry(data []byte) {
 			}
 		}
 
-		// if the keys and content fields are missing, skip the message as it
-		// does not have parsable data used by Telegraf
-		if keys == nil || content == nil {
+		if content == nil && !c.IncludeDeleteField {
+			c.Log.Debugf("Message skipped because no content found and include of delete field not enabled")
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go:391]:
```suggestion
			c.Log.Debug("Message skipped because no content found and include of delete field not enabled")
```

**Review by srebhan — COMMENTED** [review]
One small improvement and please address my comments for `sample.conf`.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go (hunk 14 above)
Comment (severindellsperger) [reply plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt.go]: Thanks for pointing that out.

File: plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go [hunk 15]
```diff
@@ -96,6 +100,176 @@ func TestHandleTelemetryTwoSimple(t *testing.T) {
...
+		acc := &testutil.Accumulator{}
+		// error is expected since we are passing in dummy transport
+		require.ErrorContains(t, c.Start(acc), "dummy")
+		data, err := proto.Marshal(test.telemetry)
+		require.NoError(t, err)
+
+		c.handleTelemetry(data)
+		actual := acc.GetTelegrafMetrics()
+		fmt.Println(test.expected)
+		fmt.Println(actual)
```
Comment (srebhan) [inline plugins/inputs/cisco_telemetry_mdt/cisco_telemetry_mdt_test.go:268]: Can those be removed please
```suggestion
```

Furthermore, please resolve the merge conflict.

**Comment by srebhan** [comment]
Oh and you should run `make docs`...

**Review by srebhan — APPROVED** [review]
Looks good to me. Thanks for keeping this going @severindellsperger!

**Review by powersj — APPROVED** [review]
Huge thank you severindellsperger for driving this here and with cisco!
